source/common/conn_pool/conn_pool_base.cc

```diff
@@ -517,10 +565,12 @@ void ConnPoolImplBase::onConnectionEvent(ActiveClient& client, absl::string_view
     client.has_handshake_completed_ = true;
     client.conn_connect_ms_->complete();
     client.conn_connect_ms_.reset();
-    ASSERT(client.state() == ActiveClient::State::Connecting);
-    bool streams_available = client.currentUnusedCapacity() > 0;
-    transitionActiveClientState(client, streams_available ? ActiveClient::State::Ready
-                                                          : ActiveClient::State::Busy);
+    if (client.state() == ActiveClient::State::Connecting ||
+        client.state() == ActiveClient::State::ReadyForEarlyData) {
```

Comment:
Are these the only two possible states? If so, would it make sense to ASSERT this and avoid the conditional?

Reply:
Nope, early data ready clients can become busy or draining, in which case their state shouldn't be changed if they get Connected.